src/config/pragmas/plugins.js

```diff
@@ -1,9 +1,114 @@
+  let pluginItems = []
+  let tagPlugins = (arr, type) => pluginItems.push(...arr.map(p => ({ plugin: p, type })))
+  if (arc?.plugins?.length) tagPlugins(arc.plugins, 'plugin')
+  if (arc?.macros?.length) tagPlugins(arc.macros, 'macro')
 
+  for (let pluginItem of pluginItems) {
+    let { plugin, type } = pluginItem
+    let name
+    let pluginPath
+
+    if (is.string(plugin)) {
+      name = plugin
+      pluginPath = getPath(cwd, type + 's', name)
+    }
+    else if (is.object(plugin)) {
+      name = Object.keys(plugin)[0]
+      pluginPath = plugin[name].src
+        ? normalizeSrc(cwd, plugin[name].src)
+        : join(cwd, 'src', type + 's', name)
+    }
+
+    if (reservedNames.includes(name)) {
+      errors.push(`Plugin name ${name} is reserved, please rename your plugin`)
+      continue
+    }
+    if (!validationPatterns.veryLooseName.test(name)) {
+      errors.push('Plugin names can only contain [a-zA-Z0-9/\\-._]')
+      continue
+    }
+    if (pluginPath) {
+      try {
+        if (type === 'plugin') {
+          // eslint-disable-next-line
+          plugins[name] = require(pluginPath)
+        }
+        // Remap @macros to deploy.start
+        if (type === 'macro') {
+          // eslint-disable-next-line
+          plugins[name] = { deploy: { start: require(pluginPath) } }
+        }
+        // Walk each plugin and build the method tree
+        Object.entries(plugins[name]).forEach(([ method, item ]) => {
+          // Primitive setters
+          if (method === 'set') {
+            if (!methods.set) methods.set = {}
+            setters.forEach(setter => {
+              if (item[setter]) {
+                let fn = item[setter]
+                if (!is.fn(fn) || fn.constructor.name === 'AsyncFunction') {
+                  let msg = `Invalid plugin, setters must be synchronous functions: plugin: ${name}, method: set.${setter}`
+                  errors.push(msg)
+                  return
+                }
+                if (!methods.set[setter]) methods.set[setter] = []
+                fn.plugin = name
+                fn.type = type
+                methods.set[setter].push(fn)
+              }
+            })
+          }
+          // Command hooks
+          else if (pluginMethods.includes(method)) {
+            Object.entries(item).forEach(([ hook, fn ]) => {
+              if (!is.fn(fn)) {
+                let msg = `Invalid plugin, must be a function: plugin: ${name}, method: ${method}.${hook}`
+                errors.push(msg)
+                return
+              }
+              if (!methods[method]) methods[method] = {}
+              if (!methods[method][hook]) methods[method][hook] = []
+              fn.plugin = name
+              fn.type = type
+              methods[method][hook].push(fn)
+            })
+          }
+        })
+      }
+      catch (err) {
+        errors.push(`Unable to load plugin '${name}': ${err.message.split('\n')[0]}`)
```

Comment:
I was playing around with my plugin and this one threw me for a loop for a while. If you have a syntax error in your plugin code, just reporting the first line of the error as is done here makes it tough to figure out the problem as the stack trace from the plugin is removed. Example error:

```
➜ arc sandbox
[
  "Unable to load plugin 'arc-plugin-s3-image-bucket': arc is not defined"
]
sandbox failed! Plugin error:
- Unable to load plugin 'arc-plugin-s3-image-bucket': arc is not defined
Error: Plugin error:
- Unable to load plugin 'arc-plugin-s3-image-bucket': arc is not defined
    at format (/Users/filmaj/src/imgbukplugapp/node_modules/@architect/inventory/src/lib/error-fmt.js:15:13)
    at architectInventory (/Users/filmaj/src/imgbukplugapp/node_modules/@architect/inventory/src/index.js:75:14)
    at main (/Users/filmaj/src/imgbukplugapp/node_modules/@architect/architect/src/index.js:72:29)
    at /Users/filmaj/src/imgbukplugapp/node_modules/@architect/architect/src/index.js:91:11
    at Object.<anonymous> (/Users/filmaj/src/imgbukplugapp/node_modules/@architect/architect/src/index.js:92:5)
    at Module._compile (internal/modules/cjs/loader.js:1063:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
    at Module.load (internal/modules/cjs/loader.js:928:32)
    at Function.Module._load (internal/modules/cjs/loader.js:769:14)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:72:12)
```

The stack trace above comes from arc/inventory, and indeed I did not have `arc` defined in the right place in my plugin code, but where in my plugin code is not reported, making fixing the problem really hard. I would reconsider how we report userland errors and stack traces in this situation.

Reply:
This is spot on! Will fix asap.

Reply:
Patched in https://github.com/architect/inventory/pull/38/commits/5825ea2e6f259c228fea16282a3bd020fc6d6660